src/Moq/Expectation.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2007, Clarius Consulting, Manas Technology Solutions, InSTEDD, and Contributors.
+// All rights reserved. Licensed under the BSD 3-Clause License; see License.txt.
+
+using System;
+using System.Linq.Expressions;
+
+using Moq.Async;
+
+namespace Moq
+{
+	/// <summary>
+	///   Represents a set (or the "shape") of invocations
+	///   against which concrete <see cref="Invocation"/>s can be matched.
+	/// </summary>
+	internal abstract class Expectation : IEquatable<Expectation>
+	{
+		public abstract LambdaExpression Expression { get; }
+
+		public virtual bool HasResultExpression(out IAwaitableFactory awaitableFactory)
+		{
+			awaitableFactory = null;
+			return false;
+		}
```

Comment:
This method doesn't really fit in very well here, as it doesn't have much to do with the core purpose of `Expectation`.

A possibly better design might be to create a custom LINQ expression tree node type `ResultExpression` (which would carry the `IAwaitableFactory` as data), then add such a node directly to `Expression`.

That can be done as a separate refactoring another time.